src/composer.ts

```diff
@@ -302,16 +314,45 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
      * @param trigger The text to look for
      * @param middleware The middleware to register
      */
-    hears(
-        trigger: MaybeArray<string | RegExp>,
-        ...middleware: Array<Middleware<HearsContext<C>>>
-    ): Composer<HearsContext<C>> {
-        const trg = triggerFn(trigger);
+    hears<
+        T extends string = string,
+        U extends FilteredMatchContext<C, T, ":text" | ":caption"> =
```

Comment:
I don't think this is correct. Now, you can pass arbitrary context types that fulfill the `extends` constraint. The result would promise to generate the necessary predicate for that, which it does not.